### PR TITLE
Respect queryParameter updates in catalog pickers

### LINKED ISSUE #8432: Catalog does not listen to change events in `location`
Issue mentioned in following PR and Discord:
PR: https://github.com/backstage/backstage/pull/8378
Discord: https://discord.com/channels/687207715902193673/687235481154617364/917756886491754526
## Expected Behavior
I should be able to use a `<Link /> ` to navigate to the catalog and use the search query to apply filters e.g.,
```
<Link to="/catalog?filters%5Bkind%5D=component&filters%5Btype%5D=service" />
```

This would allow us to have a `<SidebarSubmenuItem/>` that can be clicked to navigate to the catalog with different filters applied. 

```
          <SidebarSubmenuItem
            title="Services"
            to="catalog?filters%5Bkind%5D=component&filters%5Btype%5D=service"
            icon={HomeIcon}
          />
          <SidebarSubmenuItem
            title="Libraries"
            to="catalog?filters%5Bkind%5D=component&filters%5Btype%5D=library"
            icon={HomeIcon}
          />
```
Clicking the first one would would take us to the catalog, components filtered by service type.
Clicking the second one would change the filter from 'service' -> 'library' in the catalog (causing a re-render).

## Current Behavior

Currently, the Catalog does not react to change events in the `location`. Once it gets loaded, filtering is only done by js.
In the `<SidebarSubmenu/>` case above, 
- Clicking the first one would would take us to the catalog, components filtered by service type.
- Clicking the second one would cause a change in the location (`catalog?filters%5Bkind%5D=component&filters%5Btype%5D=service` -> `catalog?filters%5Bkind%5D=component&filters%5Btype%5D=library`) but the view won't change - the catalog doesn't re-render and continues displaying only the components with `type=service`

Vice versa: if I filter the catalog from within, `location` (`const location = useLocation()`) does not get updated, so the items in the sidebar/submenu won't be set to be active.

## Steps to Reproduce

You can try replacing the Sidebar content and replacing with this snippet locally in `Root.tsx`
```
<Sidebar disableExpandOnHover>
  <SidebarItem icon={HomeIcon} text="Catalog">
    <SidebarSubmenu title="Catalog">
              <SidebarSubmenuItem
                title="Services"
                to="catalog?filters%5Bkind%5D=component&filters%5Btype%5D=service"
                icon={HomeIcon}
              />
              <SidebarSubmenuItem
                title="Libraries"
                to="catalog?filters%5Bkind%5D=component&filters%5Btype%5D=library"
                icon={HomeIcon}
              />
    </SidebarSubmenu>
  </SidebarItem>
</Sidebar>
```

1. Go to `http://localhost:3000/1`
2. Click the 'Services' submenu item in the sidebar - this should take you to the catalog, filtered by `type=service`
3. Click the 'Libraries' submenu item in the sidebar - this will change the search query but the catalog will remain filtered by 'type=service'

Vice versa:
1. Go to http://localhost:3000/catalog
2. Manually filter by Service type - this will update search to something like `catalog?filters%5Bkind%5D=component&filters%5Btype%5D=service` .
3. The 'Service' submenu item won't be highlighted/set to active. 

## Context

Implemented new features for the Sidebar based on this issue: https://github.com/backstage/backstage/issues/6933
 PR: https://github.com/backstage/backstage/pull/7521
The design includes 3 submenu items which allow us to click on them to go to the catalog with different filters applied such as 'service', 'library' and 'website' which is not possible at the moment.

## Your Environment

- NodeJS Version (v14):
- Operating System and Version (e.g. Ubuntu 14.04):
- Browser Information:

### PR DESCRIPTION
**Pull request by timbonicus**
Closes #8432.

Previously, the query parameters were fetched once off the URL on page load and thereafter the query parameters were disregarded, only pushed one way from filter changes triggered from UI actions.

This updates `useEntityListProvider` to leverage react-router's `useLocation` hook to be aware of external updates to query parameters. The catalog pickers are updated to pay attention to the exported `queryParameters` from the hook, which previously were only used for initialization.

Loops in state are avoided by the use of `replaceState` in `useEntityListProvider`.

@danielsamet I tried to push these changes to #8763 but didn't have permission 😅  This should replace that PR. I'll look to add tests to this PR.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx [hunk 1]
```diff
@@ -169,11 +169,16 @@ export const UserListPicker = ({
     ownedFilter.filterEntity(entity),
```
Comment (mtlewis) [inline plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:169]: Entity ownership filters from query parameters are currently totally clobbered by this component until https://github.com/backstage/backstage/pull/8937 is merged - maybe worth waiting until after that's in to test the changes in this component?

Comment (mtlewis) [reply plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:169]: Merged now so I’ll resolve this comment!

**Comment by benjdlambert** [comment]
@timbonicus have you had a chance to check out what's going on with the loops?

**Comment by timbonicus** [comment]
@benjdlambert I think it's just a test issue but troubleshooting now

**Review by benjdlambert — APPROVED** [review]

**Comment by benjdlambert** [comment]
Woop woop thanks @timbonicus!

**Comment by awanlin** [comment]
Thanks @timbonicus

**Pull request merged**